tests/stream.rs

```diff
@@ -0,0 +1,51 @@
+use async_std::prelude::*;
+use async_std::stream;
+use async_std::sync::channel;
+use async_std::task;
+
+#[test]
+/// Checks that streams are merged fully even if one of the components
+/// experiences delay.
+fn merging_delayed_streams_work() {
+    let (sender, receiver) = channel::<i32>(10);
+
+    // Delay the left stream.
+    let s1 = receiver.chain(stream::once(92));
+    let s2 = stream::empty();
+    let mut s = s1.merge(s2);
+    let t1 = task::spawn(async move {
+        let mut xs = Vec::new();
+        while let Some(x) = s.next().await {
+            xs.push(x);
+        }
+        xs
+    });
+
+    task::block_on(async move {
+        task::sleep(std::time::Duration::from_millis(5)).await;
```

Comment:
Does this sleep have any significance? Perhaps it'd be safer to pick a larger duration, maybe 500ms?

Reply:
without this sleep, the test can pass by accident, as we would not poll a `Poll::Pending` out of the channel. Basically, we need to ensure that the `t` tasks is polled before we do a send via channel. Bumped the dealay to 500ms: we can trivially dial it back if we find test-suite to be too long!

Reply:
to be more precise, the unfortunate timing here can't lead to false positive: if the code is correct, the test will always pass. However, timing can lead to a false-negative, where a buggy code would pass the test. 

I've considered making this deterministic by writing a stupid `Stream` which answers with `Pending` the first time it is polled and with `Ready(Some(92))` the second time. In the end, I've decided to go with channels + sleep to simulate a similar effect: writing a stream impl manually can easily introduce errors of it's own.